src/__stories__/FormInput/index.tsx

```diff
@@ -0,0 +1,45 @@
+import * as React from 'react';
+import * as yup from 'yup';
+
+import { withKnobs } from '@storybook/addon-knobs';
+import { storiesOf } from '@storybook/react';
+
+import { FormInput } from '../../';
+
+storiesOf('FormInput', module)
+  .addDecorator(withKnobs)
+  .add('with synchronous validation', () => (
+    <div className="p-40">
+      <FormInputContainer schema={exampleSyncSchema} />
+    </div>
+  ))
+  .add('with async validation', () => (
+    <div className="p-40">
+      <FormInputContainer schema={exampleAsyncSchema} />
+    </div>
+  ));
+
+const FormInputContainer: React.FC<{ schema: yup.StringSchema }> = ({ schema }) => {
+  const [value, setValue] = React.useState('');
+
+  return (
+    <FormInput
+      value={value}
+      onChange={(e: React.ChangeEvent<HTMLInputElement>) => setValue(e.currentTarget.value)}
+      schema={schema}
+    />
+  );
+};
+
+const exampleSyncSchema = yup
+  .string()
+  .min(3, 'waaay too short')
+  .max(5, 'waaay too long');
+
+const exampleAsyncSchema = yup.string().test('oddchars', 'The number of characters must be odd', value => {
+  return new Promise(resolve => {
+    setTimeout(() => {
+      resolve(value.length % 2 === 1);
```

Comment:
What happens if the promise gets rejected?

Reply:
By the docs you are not supposed to provide anything that can reject:
>  To make a test async return a promise that resolves true or false or a ValidationError.
https://github.com/jquense/yup#mixedtestname-string-message-string--function-test-function-schema

But I haven't tried, should I?

Reply:
Hmm,  if that's the case, I am wondering wether we shouldn't catch the promise error here and resolve it with a `false`

Reply:
This is a story and an example timeout promise. What do you mean by `here`? In the hook?

Reply:
Yes, in the hook. What do you think?

Reply:
I checked the behavior now.
`yup` really does not catch the error, so it falls through to our catch handler:
```typescript
.catch(e => {
  setErrors(e.errors || unknownError);
});
```

This usually ends up as "Invalid input" displayed in the UI.

In the hook I don't have access to the actual validation promise (in a sense that I could catch the error and turn it into a regular `false`.)

Nether do I think it would be good practice on our part to override how `yup` works. The user is creating `yup` schemas, they should create well behaving ones.

What we *can* do is to make sure that our hook handles these unknown errors, and I just did that: it does. One thing to consider from an UX perspective is to rename the "Invalid input" fallback to something like "Unknown error", that conveys that it's possibly a programming error, not something on the user's side.

Reply:
Done